rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,991 @@
+
+#### Unequal and incompatible overlapping constraints:
+
+Finally, say that `Person` defines that `name` is a _string_ and is **required**, and `Superhero` defines that `name` is _an array of strings_ and is **required**.
+
+When applied simultaneously, these constraints are _incompatible_ and result in an unsatisfiable type. Let us try and create an instance that satisfies it. First we try and create one with a string `name`:
+
+```json5
+{
+  name: "John Doe", // ERROR: expected string array, got string
+  ... // superpower, age, occupation
+}
+```
+
+And then we instead try and create one with an array of `name`s:
+
+```json5
+{
+  name: ["John Doe"], // ERROR: expected string, got string array
+  ... // superpower, age, occupation
+}
+```
+
+### An uncomfortable compromise
+
+As alluded to above, checking compatibility of any two given schemas is quite possibly an NP-Hard problem. Even if it is generally solvable, it is likely a task that is complicated enough that it would be an exceptionally high barrier to entry for potential implementers to overcome.
+
+While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we opt not to encode such restrictions within the type system, and **implementations should not assume that an arbitrary type is satisfiable**.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
+- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
+- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.
+  - This may be especially difficult if a property type is defined as being an array of `oneOf` a given set of property type objects. Detecting which sub-schema applies to a given array element when accounting for dropped properties, **may** be a very difficult task.
```

Comment:
I don't really get what you mean by "projecting/selecting". If `Employee` is a subtype of `Person` and for example and `Person` has a `oneOf` property, what is the difficulty to use `Employee`? Let's assume an application, which requires a `Person`. If an `Employee` is passed to the application, the type still satisfies `Person` so the application can _just use_ the provided type. The application should only rely on the `Person`'s properties to exist.
If the `oneOf` is an array, either all properties are valid as the array is of a specific type, which satisfies `Person` or the types are not compatible.

I don't think, this is an issue.

Reply:
```suggestion
```

Reply:
> the type still satisfies Person so the application can just use the provided type

This is the assumption I think we're disagreeing on. We're not only interested in duck-typing (or similar patterns). Entity types behave different to a `type` or `interface` in TypeScript, they do not say "the object has to _at least_ have X" they define the **complete** set of possible entities.

This is the point above about needing a "projection" or "selection" of properties. If a block asks for a `Person` and you pass in `Employee`, that employee entity will _not_ satisfy the `Person` schema because we do not allow `additionalProperties: true` (or `unevaluatedProperties`, hence the section on that).

This has even more run on affects we need to consider, such as the `entityTypeId` of a 'projected' entity. The main issues though I think come about from use-cases that deal with `create` or `update` methods for entities. If they receive a subtype of the other type and `update` it, but are only aware of _some_ of the properties, then they might accidentally invalidate the subtype. For an example, let's think you have a component that lets you duplicate a `Person` entity (not to be confused with "duplication" of types outlined in this post). The component might let you pick from a list of `Person` entities, pick one to copy the properties over from, and then edit it. This list is going to contain `Employee` entities as well, as they're a subtype of `Person`. But now you try and copy the properties of one of the `Employee` entities. If you literally did `deepcopy(person.properties)` you'd get the additional properties from `Employee` and it'd break validation when you try and `createPerson`.

Reply:
> If we allow setting additionalProperties in entity types, it's possible to mark an entity as "final" - an entity type that cannot be inherited from. This would be possible by adding additionalProperties: { type: boolean } to properties on the entity type meta schema. As we currently do not enforce additionalProperties: false on entity types I think this is good for allowing applications to treat the entity types as they need to.

How exactly would that be reflected in the graph itself? I get your POV, but from my perspective, this is full of potential problems when trying to build on-top of these assumptions.

What I mean is the following: if I have a type `Person` with properties: `firstName`, and `lastName`, if we're going to assume `additionalProperties: true`, then it's valid to also have `middleName` in there (also when creating the type). What will happen in the case of, e.g. migration tools if someone now wants to add `middleName` as a "real" property?

I think this is then also ripe for abuse, where people implicitly expect (even if not in the schema) that a specific property exists (because the developer just adds it to their entities and then uses it in their blocks).

Reply:
We assume in the version-RFC an implicit `additionalProperties: false` to ensure, that entities do not have random properties, but this does not work with subtyping (as mentioned by @Alfred-Mountfield above)

Reply:
> The same applies to create and update: If the application does not correctly creates or updates an entity it's the application's fault.

I had an example that I wanted to walkthrough to discuss this, but I think it's a distraction, I think the discussion here is actually about the need for a projection mechanism.

We are discussing the need for a 'projection' here to actually satisfy the subtype property of our inheritance mechanism. It is important that a block is able to request a `Person`, and then be able to validate it against that `Person` type. If it is of type `Employee`, we want to make sure the additional properties in it are adequately captured by `Employee`, we don't want to _**also**_ allow any other random property like "potato: 32" which isn't even associated with a property type.

Reply:
@Alfred-Mountfield has perfectly summarized what I was trying to say about `additionalProperties`.

I'd suggest that instead of "we suggest the constraint is implied at the root level schema when validating data against a specific entity type" we strengthen the wording to: "the constraint **must** be implied at the root level schema when validating data against a specific entity type". Otherwise, it might lead to a split in the ecosystem if some implementation decide not to do so.

Reply:
> Is literally describing duck-typing. This is how TypeScript interfaces and types work.

I assumed another location where the duck typing happens. You don’t need to duck-type on your application as you expect the entity to be a specific entity type. Validating a JSON based on a given JSON schema is always duck typing. We don't expect the blocks to check, if `name` exists on a `Person`, it can assume it's there, but the application may need to make sure that only `Person`s are passed to the blocks.

---

However, I understand your point now and we indeed talked passed each other. As to me, it was not clear that by "projecting" you mean "coercion", which is probably the more common term when talking about subtyping and variance. For general JSON data to be passed as subtype, coercion _should_ be fairly easy but as outlined in the code it's not defined how this would happen

To conclude this thread: Do you think it's required to resolve this question prior to merging the RFC?